Segmentation violation in several CLI commands

### Describe the bug

Running `kraft fetch`, `kraft prepare`, `kraft menu` or `kraft clean` inside the `app-helloworld` repo yields the following segfault:

```console
$ kraft fetch
[ W ] could not initialize package manager: could not determine handler [format]=oci
[ W ] could not read target config file: could not open file: open /home/acotten/git/unikraft/app-helloworld/.config.helloworld_linuxu-x86_64: no such file or directory
[ W ] could not read target config file: could not open file: open /home/acotten/git/unikraft/app-helloworld/.config.helloworld_kvm-x86_64: no such file or directory
[ W ] could not read target config file: could not open file: open /home/acotten/git/unikraft/app-helloworld/.config.helloworld_xen-x86_64: no such file or directory
[ W ] could not read target config file: could not open file: open /home/acotten/git/unikraft/app-helloworld/.config.helloworld_linuxu-arm64: no such file or directory
[ W ] could not read target config file: could not open file: open /home/acotten/git/unikraft/app-helloworld/.config.helloworld_kvm-arm64: no such file or directory
[ W ] could not read target config file: could not open file: open /home/acotten/git/unikraft/app-helloworld/.config.helloworld_linuxu-arm: no such file or directory
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0xfb8536]

goroutine 1 [running]:
kraftkit.sh/unikraft/app.application.MakeArgs({{0xc00033b1f0, 0xa}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0xc0000460c4, 0x29}, ...}, ...)
        /home/acotten/git/unikraft/kraftkit/unikraft/app/application.go:301 +0x796
kraftkit.sh/unikraft/app.application.Make({{0xc00033b1f0, 0xa}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0xc0000460c4, 0x29}, ...}, ...)
        /home/acotten/git/unikraft/kraftkit/unikraft/app/application.go:317 +0x20a
kraftkit.sh/unikraft/app.application.Fetch({{0xc00033b1f0, 0xa}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0xc0000460c4, 0x29}, ...}, ...)
        /home/acotten/git/unikraft/kraftkit/unikraft/app/application.go:434 +0x1a5
kraftkit.sh/cmd/kraft/fetch.(*Fetch).Run(0x0?, 0x0?, {0x2732a80?, 0x0?, 0x0?})
        /home/acotten/git/unikraft/kraftkit/cmd/kraft/fetch/fetch.go:67 +0xda
kraftkit.sh/cmdfactory.bind.func1(0xc000004f00?, {0x2732a80, 0x0, 0x0})
        /home/acotten/git/unikraft/kraftkit/cmdfactory/builder.go:431 +0x19d
github.com/spf13/cobra.(*Command).execute(0xc000004f00, {0x2732a80, 0x0, 0x0})
        /home/acotten/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:916 +0x862
github.com/spf13/cobra.(*Command).ExecuteC(0xc000004600)
        /home/acotten/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:1044 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
        /home/acotten/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:968
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        /home/acotten/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:961
kraftkit.sh/cmdfactory.Main({0x1d09688?, 0xc000542150}, 0xc000004600)
        /home/acotten/git/unikraft/kraftkit/cmdfactory/builder.go:103 +0x73
main.main()
        /home/acotten/git/unikraft/kraftkit/cmd/kraft/kraft.go:142 +0x3a5
```

### Steps to reproduce

_No response_

### Expected behavior

_No response_

### Which architectures were you using or does this bug affect?

_No response_

### Which operating system were you using or does this bug affect?

_No response_

### Relevant log output

_No response_

--- a/cmd/kraft/build/build.go
+++ b/cmd/kraft/build/build.go
@@ -16,6 +16,7 @@ import (
 	"kraftkit.sh/cmdfactory"
 	"kraftkit.sh/config"
 	"kraftkit.sh/exec"
+	"kraftkit.sh/internal/cli"
 	"kraftkit.sh/pack"
 	"kraftkit.sh/unikraft"
 
@@ -339,43 +340,13 @@ func (opts *Build) Run(cmd *cobra.Command, args []string) error {
 
 	processes = []*paraprogress.Process{} // reset
 
-	var selected target.Targets
-
-	// Filter the targets by CLI selection
-	for _, targ := range project.Targets() {
-		switch true {
-		case
-			// If no arguments are supplied
-			len(opts.Target) == 0 &&
-				len(opts.Architecture) == 0 &&
-				len(opts.Platform) == 0,
-
-			// If the --target flag is supplied and the target name match
-			len(opts.Target) > 0 &&
-				targ.Name() == opts.Target,
-
-			// If only the --arch flag is supplied and the target's arch matches
-			len(opts.Architecture) > 0 &&
-				len(opts.Platform) == 0 &&
-				targ.Architecture().Name() == opts.Architecture,
-
-			// If only the --plat flag is supplied and the target's platform matches
-			len(opts.Platform) > 0 &&
-				len(opts.Architecture) == 0 &&
-				targ.Platform().Name() == opts.Platform,
-
-			// If both the --arch and --plat flag are supplied and match the target
-			len(opts.Platform) > 0 &&
-				len(opts.Architecture) > 0 &&
-				targ.Architecture().Name() == opts.Architecture &&
-				targ.Platform().Name() == opts.Platform:
-
-			selected = append(selected, targ)
-
-		default:
-			continue
-		}
-	}
+	// Filter project targets by any provided CLI options
+	selected := cli.FilterTargets(
+		project.Targets(),
+		opts.Architecture,
+		opts.Platform,
+		opts.Target,
+	)
 
 	if len(selected) == 0 {
 		return fmt.Errorf("no targets selected to build")

--- a/cmd/kraft/prepare/prepare.go
+++ b/cmd/kraft/prepare/prepare.go
@@ -11,11 +11,16 @@ import (
 	"github.com/spf13/cobra"
 
 	"kraftkit.sh/cmdfactory"
+	"kraftkit.sh/internal/cli"
 	"kraftkit.sh/packmanager"
 	"kraftkit.sh/unikraft/app"
 )
 
-type Prepare struct{}
+type Prepare struct {
+	Architecture string `long:"arch" short:"m" usage:"Filter prepare based on a target's architecture"`
+	Platform     string `long:"plat" short:"p" usage:"Filter prepare based on a target's platform"`
+	Target       string `long:"target" short:"t" usage:"Filter prepare based on a specific target"`
+}
 
 func New() *cobra.Command {
 	return cmdfactory.New(&Prepare{}, cobra.Command{
@@ -72,6 +77,18 @@ func (opts *Prepare) Run(cmd *cobra.Command, args []string) error {
 	)
 	if err != nil {
 		return err
+	}
+
+	// Filter project targets by any provided CLI options
+	selected := cli.FilterTargets(
+		project.Targets(),
+		opts.Architecture,
+		opts.Platform,
+		opts.Target,
+	)
+
+	if len(selected) == 1 {
+		return project.Prepare(ctx, selected[0])
 	}
 
 	return project.Prepare(ctx, nil)

--- a/cmd/kraft/run/run.go
+++ b/cmd/kraft/run/run.go
@@ -9,10 +9,8 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
-	"sort"
 
 	"github.com/MakeNowJust/heredoc"
-	"github.com/erikgeiser/promptkit/selection"
 	"github.com/moby/moby/pkg/namesgenerator"
 	"github.com/rancher/wrangler/pkg/signals"
 	"github.com/sirupsen/logrus"
@@ -21,6 +19,7 @@ import (
 	"kraftkit.sh/cmdfactory"
 	"kraftkit.sh/config"
 	"kraftkit.sh/exec"
+	"kraftkit.sh/internal/cli"
 	"kraftkit.sh/iostreams"
 	"kraftkit.sh/log"
 	"kraftkit.sh/machine"
@@ -32,7 +31,6 @@ import (
 	"kraftkit.sh/unikraft"
 	"kraftkit.sh/unikraft/app"
 	"kraftkit.sh/unikraft/target"
-	"kraftkit.sh/utils"
 )
 
 type Run struct {
@@ -302,7 +300,6 @@ func (opts *Run) Run(cmd *cobra.Command, args []string) error {
 		)
 		// d). use a defined working directory as a Unikraft project
 	} else if len(workdir) > 0 {
-		targetName := opts.Target
 		project, err := app.NewProjectFromOptions(
 			ctx,
 			app.WithProjectWorkdir(workdir),
@@ -312,42 +309,31 @@ func (opts *Run) Run(cmd *cobra.Command, args []string) error {
 			return err
 		}
 
-		if len(project.TargetNames()) == 1 {
-			targetName = project.TargetNames()[0]
-			if len(opts.Target) > 0 && opts.Target != targetName {
-				return fmt.Errorf("unknown target: %s", opts.Target)
-			}
+		// Filter project targets by any provided CLI options
+		targets := cli.FilterTargets(
+			project.Targets(),
+			opts.Architecture,
+			opts.Platform,
+			opts.Target,
+		)
 
-		} else if targetName == "" && len(project.Targets()) > 1 {
-			if config.G[config.KraftKit](ctx).NoPrompt {
-				return fmt.Errorf("with 'no prompt' enabled please select a target")
-			}
+		var t target.Target
 
-			sp := selection.New("select target:", selectionTargets(project.Targets()))
-			sp.Filter = nil
+		switch {
+		case len(targets) == 1:
+			t = targets[0]
 
-			selectedTarget, err := sp.RunPrompt()
+		case config.G[config.KraftKit](ctx).NoPrompt && len(targets) > 1:
+			return fmt.Errorf("could not determine what to run based on provided CLI arguments")
+
+		case config.G[config.KraftKit](ctx).NoPrompt && len(targets) == 0:
+			return fmt.Errorf("could not match any target")
+
+		default:
+			t, err = cli.SelectTarget(targets)
 			if err != nil {
 				return err
 			}
-
-			targetName = selectedTarget
-
-		} else if targetName != "" && utils.Contains(project.TargetNames(), targetName) {
-			return fmt.Errorf("unknown target: %s", targetName)
-		}
-
-		t, err := project.TargetByName(targetName)
-		if err != nil {
-			return err
-		}
-
-		// Validate selection of target
-		if len(opts.Architecture) > 0 && (opts.Architecture != t.Architecture().Name()) {
-			return fmt.Errorf("selected target (%s) does not match specified architecture (%s)", target.TargetPlatArchName(t), opts.Architecture)
-		}
-		if len(opts.Platform) > 0 && (opts.Platform != t.Platform().Name()) {
-			return fmt.Errorf("selected target (%s) does not match specified platform (%s)", target.TargetPlatArchName(t), opts.Platform)
 		}
 
 		name := t.Name()
@@ -483,17 +469,4 @@ func (opts *Run) Run(cmd *cobra.Command, args []string) error {
 	}
 
 	return nil
-}
-
-// selectionTargets returns the given target.Targets in a format suitable for
-// interactive prompts.
-func selectionTargets(tgts target.Targets) []string {
-	tgtStrings := make([]string, 0, len(tgts))
-	for _, tgt := range tgts {
-		tgtStrings = append(tgtStrings, fmt.Sprintf("%s (%s)", tgt.Name(), target.TargetPlatArchName(tgt)))
-	}
-
-	sort.Strings(tgtStrings)
-
-	return tgtStrings
 }

--- a/internal/cli/target.go
+++ b/internal/cli/target.go
@@ -1,0 +1,92 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2022, Unikraft GmbH and The KraftKit Authors.
+// Licensed under the BSD-3-Clause License (the "License").
+// You may not use this file except in compliance with the License.
+package cli
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/erikgeiser/promptkit/selection"
+	"kraftkit.sh/unikraft/target"
+)
+
+// SelectTarget is a utility method used in a CLI context to prompt the user
+// for a specific application's target.
+func SelectTarget(targets target.Targets) (target.Target, error) {
+	names := make([]string, 0, len(targets))
+	for _, t := range targets {
+		names = append(names, fmt.Sprintf(
+			"%s (%s)",
+			t.Name(),
+			target.TargetPlatArchName(t),
+		))
+	}
+
+	sort.Strings(names)
+
+	sp := selection.New("select target:", names)
+	sp.Filter = nil
+
+	result, err := sp.RunPrompt()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, t := range targets {
+		if fmt.Sprintf("%s (%s)", t.Name(), target.TargetPlatArchName(t)) == result {
+			return t, nil
+		}
+	}
+
+	return nil, fmt.Errorf("could not select target")
+}
+
+// FilterTargets returns a subset of `targets` based in input strings `arch`,
+// `plat` and/or `targ`
+func FilterTargets(targets target.Targets, arch, plat, targ string) target.Targets {
+	var selected target.Targets
+
+	type condition func(target.Target, string, string, string) bool
+
+	conditions := []condition{
+		// If no arguments are supplied
+		func(t target.Target, arch, plat, targ string) bool {
+			return len(targ) == 0 && len(arch) == 0 && len(plat) == 0
+		},
+
+		// If the `targ` is supplied and the target name match
+		func(t target.Target, arch, plat, targ string) bool {
+			return len(targ) > 0 && t.Name() == targ
+		},
+
+		// If only `arch` is supplied and the target's arch matches
+		func(t target.Target, arch, plat, targ string) bool {
+			return len(arch) > 0 && len(plat) == 0 && t.Architecture().Name() == arch
+		},
+
+		// If only `plat`` is supplied and the target's platform matches
+		func(t target.Target, arch, plat, targ string) bool {
+			return len(plat) > 0 && len(arch) == 0 && t.Platform().Name() == plat
+		},
+
+		// If both `arch` and `plat` are supplied and match the target
+		func(t target.Target, arch, plat, targ string) bool {
+			return len(plat) > 0 && len(arch) > 0 && t.Architecture().Name() == arch && t.Platform().Name() == plat
+		},
+	}
+
+	// Filter `targets` by input arguments `arch`, `plat` and/or `targ`
+	for _, t := range targets {
+		for _, c := range conditions {
+			if !c(t, arch, plat, targ) {
+				continue
+			}
+
+			selected = append(selected, t)
+		}
+	}
+
+	return selected
+}

--- a/unikraft/app/application.go
+++ b/unikraft/app/application.go
@@ -298,7 +298,11 @@ func (app application) MakeArgs(tc target.Target) (*core.MakeArgs, error) {
 		OutputDir:      app.outDir,
 		ApplicationDir: app.workingDir,
 		LibraryDirs:    strings.Join(libraries, core.MakeDelimeter),
-		ConfigPath:     filepath.Join(app.workingDir, tc.ConfigFilename()),
+	}
+
+	// Set the relevant Unikraft `.config` file when a target is set
+	if tc != nil {
+		args.ConfigPath = filepath.Join(app.workingDir, tc.ConfigFilename())
 	}
 
 	if tc != nil {

